tools/src/keygen/KeygenOptions.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE ApplicativeDo #-}
+
+module KeygenOptions
+    ( KeygenConfig(..)
+    , keygenConfigParser
+    ) where
+
+import Codec.Serialise (Serialise)
+import Options.Applicative (Parser, ReadM, auto, flag', help, long, metavar, option, str, strOption)
+
+import Dscp.CommonCLI
+import Dscp.Crypto
+import Dscp.Resource.Keys
+import Dscp.Resource.Logging
+import Dscp.Rio
+import Dscp.Util
+import Dscp.Util.Serialise
+
+-- | All ways to convert given text to raw data we want be able to use when
+-- parsing CLI options.
+bytestringDecoders :: [Text -> Either String ByteString]
+bytestringDecoders = [fromBase64, fromHex, pure . encodeUtf8]
+
+-- | Read something bytestringy.
+baReadM :: FromByteArray a => ReadM a
+baReadM =
+    asum $ bytestringDecoders <&> \converter -> do
+        text :: Text <- str
+        bs :: ByteString <- leftToFail $ converter text
+        leftToFail $ fromByteArray bs
+
+-- | Read something 'Serialise.serialise'd.
+serialiseReadM :: Serialise a => ReadM a
+serialiseReadM =
+    asum $ bytestringDecoders <&> \converter -> do
+        text :: Text <- str
+        bs :: ByteString <- leftToFail $ converter text
+        leftToFail @Text . first show $ deserialiseOrFail' bs
+
+-- | Parser for all possble ways to give secret key.
+secretKeyParser :: Parser (IO SecretKey)
+secretKeyParser = asum
+    [ fmap pure . option baReadM $
+          long "secret" <>
+          metavar "TEXT/BYTESTRING" <>
+          help "Use given secret key as base."
+
+    , do
+      esk <- option serialiseReadM $
+          long "secret" <>
+          metavar "TEXT/BYTESTRING" <>
+          help "Use given encrypted secret key as base."
+      pp <- optional . option passphraseReadM $
+          long "password" <>
+          metavar "TEXT" <>
+          help "Password from given key. If no password is specified, secret \
+               \key is assumed to have no password."
+      return $ pure $
+          leftToPanic @Text . first show $
+          decrypt (pp ?: emptyPassPhrase) esk
+
+    , do
+      seed <- option auto $
+          long "seed" <>
+          metavar "NUMERIC" <>
+          help "Use given seed to generate base secret."
+      return $ pure $ fst $ withIntSeed seed keyGen
+
+    , do
+      _ <- flag' () $
+          long "random" <>
+          help "Use given seed to generate base secret."
+      return $ runSecureRandom genSecretKey
+
+    , do
+      file <- strOption $
+          long "keyfile" <>
+          metavar "TEXT/BYTESTRING" <>
+          help "Use given encrypted secret key as base."
+      pp <- optional . option passphraseReadM $
+          long "password" <>
+          metavar "TEXT" <>
+          help "Password from given key. If no password is specified, secret \
+               \key is assumed to have no password."
+      return $ runRIO (noLogging @IO) $ do
+          KeyResources{..} <- readStore file (pp ?: emptyPassPhrase)
+          return _krSecretKey
+    ]
```

Comment:
This is awful, copypaste and formatting is incomprehensible, also doing some actual logic _right inside options parser_ is something I cannot really justify unless you're on the programming competition and don't have extra 10 minutes to come up with some `SecretInputType` ADT. Please make it decent.